components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/device/UserAuthenticationEndpoint.java

```diff
@@ -0,0 +1,159 @@
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.identity.oauth.endpoint.device;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.http.client.utils.URIBuilder;
+import org.wso2.carbon.identity.application.authentication.framework.model.CommonAuthRequestWrapper;
+import org.wso2.carbon.identity.core.util.IdentityUtil;
+import org.wso2.carbon.identity.oauth.cache.AppInfoCache;
+import org.wso2.carbon.identity.oauth.common.exception.InvalidOAuthClientException;
+import org.wso2.carbon.identity.oauth.dao.OAuthAppDO;
+import org.wso2.carbon.identity.oauth.endpoint.authz.OAuth2AuthzEndpoint;
+import org.wso2.carbon.identity.oauth.endpoint.exception.InvalidRequestParentException;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.device.constants.Constants;
+import org.wso2.carbon.identity.oauth2.device.dao.DeviceFlowPersistenceFactory;
+import org.wso2.carbon.identity.oauth2.device.model.DeviceFlowDO;
+import org.wso2.carbon.identity.oauth2.util.OAuth2Util;
+
+import java.io.IOException;
+import java.net.URISyntaxException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.Response;
+
+@Path("/device")
+public class UserAuthenticationEndpoint {
+
+    private static final Log log = LogFactory.getLog(UserAuthenticationEndpoint.class);
+
+    private OAuth2AuthzEndpoint oAuth2AuthzEndpoint = new OAuth2AuthzEndpoint();
+    private DeviceFlowDO deviceFlowDO = new DeviceFlowDO();
+
+    @GET
+    @Path("/")
+    @Consumes("application/x-www-form-urlencoded")
+    @Produces("text/html")
+    public Response deviceAuthorize(@Context HttpServletRequest request, @Context HttpServletResponse response)
+            throws URISyntaxException, InvalidRequestParentException, IdentityOAuth2Exception, IOException {
+
+        String userCode = request.getParameter(Constants.USER_CODE);
+        if (StringUtils.isBlank(userCode)) {
+            response.sendRedirect(IdentityUtil.getServerURL("/authenticationendpoint/device.do",
+                    false, false));
+            return null;
+        }
+
+        String clientId = DeviceFlowPersistenceFactory.getInstance().getDeviceFlowDAO().getClientIdByUserCode(userCode);
+
+        if (StringUtils.isNotBlank(clientId) && StringUtils.equals(getUserCodeStatus(userCode), Constants.PENDING)) {
+
+            setCallbackURI(clientId);
+            DeviceFlowPersistenceFactory.getInstance().getDeviceFlowDAO().setAuthenticationStatus(userCode,
+                    Constants.USED);
```

Comment:
Should we use this intermediate state?
What happens if a user enters the user code and does not authenticate?
According to the current code in this state the device will also get a pending state response.